### DISCUSSION TITLE
OpenAPI documentation

### DISCUSSION THREAD

**Discussion by cpoisnel** [pretix/pretix#2252] [Ideas] [answered]
Hi,

We work on using REST API to read, create, update and delete Events, Orders, etc.

Our main development language is Java. So at first time, we're looking at a Java client.
Before beginning a new Java client, we think about writing an [OpenAPI](https://swagger.io/specification/) documentation (ex swagger) to describe all of your API. We could after that generate a Java client (or PHP one).

Endpoints documentation could also be generated from the OpenAPI specification, with a [Sphinx extension](https://github.com/sphinx-contrib/openapi), despite the fact that your API documentation is already really clear :+1: 

Our question is simple: has anyone proposed an OpenAPI specification yet?

Best,

**Comment by raphaelm** [comment, marked as answer]
I'm not sure this was discussed before (outside of my head), no.

This is mainly a question of cost vs benefit. I'm personally not a big fan of auto-generated API clients, but I generally absolutely do see the value of an OpenAPI schema.

The main problem is that I believe auto-generated documentation is a trap. Whenever I use an API that has visibly auto-generated documentation (think your average swagger file), the documentation is **horrible**. Endpoints are ordered alphabetically or in code order, not logically by topic or feature. The examples are plain wrong most of the time, and most importantly the docs more often than not only cover **syntax** of request and responses, not **semantics**. This gets worse if you're not only auto-generating schema→docs, but also code→schema. Generating the schema from code would also be quite hard due to pretix' dynamic plugin system and some other issues (I just wanted to try [this](https://www.django-rest-framework.org/api-guide/schemas/#generating-an-openapi-schema) just for fun and it fails horribly everywhere we build serializer contexts).

So if we want to maintain the very high quality that our documentation (hopefully) has right now, we'd probably end up maintaining *both* the docs and the OpenAPI schema mostly by hand, maybe with some tooling to make it a little easier. This not only introduces potential for more errors where both documents disagree, but it also adds a significant cost to any future API change or extension.

So my personal feeling is that it's not going to be worth it. :(

**Reply by cpoisnel** [reply]
Thank you for your clear answer.
We may use the OpenAPI specification to create an error-free API client (but without auto-generated documentation).

**Reply by stapelberg** [reply]
> This is mainly a question of cost vs benefit. I'm personally not a big fan of auto-generated API clients, but I generally absolutely do see the value of an OpenAPI schema.

I have come to appreciate auto-generated REST clients more and more over time. Generating them with the OpenAPI tooling is easier and more reliable compared to writing them by hand. And, when you fix an error in the spec file, you fix all the generated clients at once. Writing clients by hand seems to be the only current option with the pretix API docs, unless I missed a machine-readable API description somewhere?

> The main problem is that I believe auto-generated documentation is a trap. Whenever I use an API that has visibly auto-generated documentation (think your average swagger file), the documentation is horrible. Endpoints are ordered alphabetically or in code order, not logically by topic or feature. The examples are plain wrong most of the time, and most importantly the docs more often than not only cover syntax of request and responses, not semantics. This gets worse if you're not only auto-generating schema→docs, but also code→schema. Generating the schema from code would also be quite hard due to pretix' dynamic plugin system and some other issues (I just wanted to try [this](https://www.django-rest-framework.org/api-guide/schemas/#generating-an-openapi-schema) just for fun and it fails horribly everywhere we build serializer contexts).

I agree that auto-generating a schema sounds like a recipe for ending up with a bad one.

As for the rest, I’d argue that’s an issue with the quality of the spec file, not with the OpenAPI documentation viewer/generators.

Directly to your points:

* editor.swagger.io respects the path order I declare in my spec file,
* the examples have a high chance of working because they are based on the data model which is then used in the API and thereby verified. 
* The description field is often required in OpenAPI, which encourages documenting the semantics.

> So if we want to maintain the very high quality that our documentation (hopefully) has right now, we'd probably end up maintaining both the docs and the OpenAPI schema mostly by hand, maybe with some tooling to make it a little easier. This not only introduces potential for more errors where both documents disagree, but it also adds a significant cost to any future API change or extension.

Having both an OpenAPI spec and a REST API documentation could also be seen as a way to discover errors with either one.

But, after writing an OpenAPI for the customers and memberships APIs, I think the current REST API docs could easily be replaced (bit by bit, I suppose) with OpenAPI-based ones. 

Here’s what my quick & dirty OpenAPI looks like:

![2022-07-24-pretix-openapi](https://user-images.githubusercontent.com/55506/180660236-60ebadb1-969a-443a-95aa-ab2e78f44d00.jpg)

Would you be open to creating a repository under the pretix org that contains a (community-contributed) OpenAPI spec? I could send what I have, and folks could add to it as they need more features of the API.

Here’s the spec file in case you want to play around with it:
[2022-07-24-pretix-openapi2.yaml.zip](https://github.com/pretix/pretix/files/9176383/2022-07-24-pretix-openapi2.yaml.zip)

**Reply by raphaelm** [reply]
Thank you for the perspective. I agree that most of my concerns would be solved if the OpenAPI file is carefully and lovingly hand-written.

Currently, every change to the API requires
- a code change
- a test change
- a docs change

Which is already often quite a lot of boilerplate, so I'm quite firm on not adding a fourth step.

However, I can certainly imagine a future where we maintain a high-value OpenAPI spec and generate the docs from there. I could even imagine a sphinx-to-openapi transpiler to start from the existing API documentation and only filling in the blanks or - probably better - an openapi-to-sphinx transpiler to replace the current documentation endpoint by endpoint as a gradual process.

I could even imagine building a tool chain into pretix that allows plugins to have their own OpenAPI spec and then auto-combining them into an instance-specific spec, which might be great for docs, but probably less great for client code generation.

After thinking about it more, this sounds quite appealing, but I'll probably not have capacity to deal with it really soon and I would need to learn more about OpenAPI to figure out if we'd lose anything important.

As for a community-maintained repo under the pretix org, I'm a little bit torn. It would certainly be nice, on the other hand it will also either remain very incomplete or put a lot of work on us (someone will need to review pull requests probably?) and we'd need to add lots of warning signs that nobody is officially maintaining it to keep up with updates of the pretix API. Not sure if that's worth it. We tried a similar thing once with the docker-compose repo which is basically defunct because we receive mainly low-quality PRs and have no time to deal with them :/

**Reply by stapelberg** [reply]
> Which is already often quite a lot of boilerplate, so I'm quite firm on not adding a fourth step.

That’s fair. Though perhaps the docs change can be *replaced* by an OpenAPI change that results in a docs change, thereby keeping the total number of steps the same. But that’s far out in the future probably. 

> […] probably better - an openapi-to-sphinx transpiler to replace the current documentation endpoint by endpoint as a gradual process.

Yeah, this sounds like the best approach to me. That way, you can write the OpenAPI and verify you got it right by replacing the existing docs parameter-for-parameter.

> As for a community-maintained repo under the pretix org, I'm a little bit torn. It would certainly be nice, on the other hand it will also either remain very incomplete or put a lot of work on us (someone will need to review pull requests probably?) and we'd need to add lots of warning signs that nobody is officially maintaining it to keep up with updates of the pretix API. 

Agreed. Luckily, the PR reviews can probably be quite quick — just open the suggested file in https://editor.swagger.io/ or similar and verify that the OpenAPI docs rendering matches the pretix API docs. And even if folks don’t get it right, it’s not a big deal for a community-maintained repository. Whoever notices any bug(s) can send a fix :) 

> Not sure if that's worth it. We tried a similar thing once with the docker-compose repo which is basically defunct because we receive mainly low-quality PRs and have no time to deal with them :/

The docker-compose repository is where I got the idea from! :)

That repository has actually helped me run a pretix disaster recovery tests in which I needed to use docker containers for every part of a pretix setup, including redis and postgres. The changes in the docker-compose repo might not be sufficient to actually use docker-compose, but they were still helpful for my purpose.

But ultimately the decision is up to you of course!

**Reply by stapelberg** [reply]
Just wanted to follow up here with an updated version of my OpenAPI description: [2023-07-31-pretix-openapi2.yaml.zip](https://github.com/pretix/pretix/files/12216806/2023-07-31-pretix-openapi2.yaml.zip) — over time, it has grown to now include requests for updating memberships and gift cards, listing orders, subevents and checkinlist positions and redeeming checkins.
